### PR TITLE
[usm] service discovery, fetch only target environment variables from /proc

### PR DESCRIPTION
**Pull request by yuri-lipnesh**
### What does this PR do?

Creates a new function to read /proc/pid/environ in chunks and save only the target environment variables.

### Motivation

A process may have a very long list of environment variables, and some of them may contain sensitive data.
Extracting the entire list of all environment variables into memory may be expensive and unsafe.
Reading the environment file chunks consumes less runtime memory. 
Extracting only the variables of interest prevents sensitive data from being loaded into memory and potentially leaked.

### Describe how to test/QA your changes

New unit tests compare results of new function to fetch environment variables to results of previously implemented function.

### Additional Notes

Benchmark comparison with previously implemented env. search function which read entire file into memory.

runtime improved by 20%
bytes allocation improved 2.2 times
number of allocations increased 2 times

```
BenchmarkGetEnvs-4                     	   73881	     16596 ns/op	   21416 B/op	      46 allocs/op
BenchmarkGetEnvsTarget-4               	   85982	     13679 ns/op	    9799 B/op	      91 allocs/op

```

### CODE REVIEW

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 1]
```diff
@@ -153,3 +186,168 @@ func getEnvs(proc *process.Process) (map[string]string, error) {
...
+	targets map[uint64]string // map of environment variables of interest
+	buffer  []byte            // buffer for reading a fragment from a file
+	end     int               // last index of data read
+	err     error             // last detected error
+	envs    map[string]string // collected environment variables
+	env     []byte            // the last processed environment variable
+}
+
+// NewEnvScanner returns a new [EnvScanner] to read from path.
+func NewEnvScanner(path string, bufSize int, targets []string) *EnvScanner {
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/envs.go:204]: Why can't we just use `bufio.Scanner`? It's already used in other places in the agent for reading env variables and should be good enough in terms of performance and will result in much simpler code. The `NewTextScannerIterator` in `pkg/security/ptracer/proc.go` is exported so it could even be reused directly.

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 2]
```diff
@@ -153,3 +186,168 @@ func getEnvs(proc *process.Process) (map[string]string, error) {
...
+			break
+		}
+	}
+
+	return nil
+}
+
+// getEnvironPath return path '/proc/<PID>/environ'
+func getEnvironPath(proc *process.Process) string {
+	hostProc := os.Getenv("HOST_PROC")
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/envs.go:317]: `kernel.HostProc()` can be used instead.

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 3]
```diff
@@ -30,6 +33,36 @@ const (
...
+	// envVarDdService - name of target environment variable DD_SERVICE
+	envVarDdService = "DD_SERVICE"
+	// envVarDdTags - name of target environment variable DD_TAGS
+	envVarDdTags = "DD_TAGS"
+	// envVarDdInjectionEnabled - name of target environment variable DD_INJECTION_ENABLED
+	envVarDdInjectionEnabled = "DD_INJECTION_ENABLED"
+	// envVarDiscoveryEnabled - name of target environment variable DD_DISCOVERY_ENABLED
+	envVarDiscoveryEnabled = "DD_DISCOVERY_ENABLED"
+	// envVarOtelServiceName - name of target environment variable OTEL_SERVICE_NAME
+	envVarOtelServiceName = "DD_OTEL_SERVICE_NAME"
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/envs.go:54]: Indirection with constants for these don't seem to add much value. The strings could be used directly instead.

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 4]
```diff
@@ -30,6 +33,36 @@ const (
...
+)
+
+// map of environment variables of interest
+var targetEnvs = []string{
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/envs.go:58]: This list should include all the environment variables needed for service name and instrumentation detection (which is why the tests are failing currently with this PR). There is an initial list in https://datadoghq.atlassian.net/browse/USMON-1126 but it should be double-checked.

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 1 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: I changed code to use bufio.Scanner

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 3 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go:52]: Without constants these strings are duplicated in functional code and in unit tests.

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 4 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go:56]: I've added some env. vars I discovered in the code and tests. However variables mentioned in the ticket USMON-1126 include variables which are not currently used. Please confirm if indeed need to include all variables in this ticket to env. extraction.

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 2 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: I've changed code to use kernel.HostProc

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 5]
```diff
@@ -153,3 +184,135 @@ func getEnvs(proc *process.Process) (map[string]string, error) {
...
+func nullTerminatedString(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	if i := bytes.IndexByte(data, 0); i >= 0 {
+		return i + 1, data[0:i], nil
+	}
+	if atEOF && len(data) > 0 {
+		// end of the file, return the remaining data
+		return len(data), data, nil
+	}
+
+	return 0, nil, nil
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/envs.go:230]: Is this implementation correct? Why does zeroSplitter in pkg/security/ptrace have different return values here? Also, can't we just reuse NewTextScannerIterator() which is exported?

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 6]
```diff
@@ -153,3 +184,135 @@ func getEnvs(proc *process.Process) (map[string]string, error) {
...
+// newEnvScanner returns a new [EnvScanner] to read from path.
+func newEnvScanner(proc *process.Process, onlyTargets bool) (*EnvScanner, error) {
+	envPath := kernel.HostProc(strconv.Itoa(int(proc.Pid)), "environ")
+	file, err := os.Open(envPath)
+	if err != nil {
+		return nil, err
+	}
+	var targetsMap map[uint64]string
+	if onlyTargets {
+		targetsMap = buildTargetsMap()
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/envs.go:242]: Why are we construction this map everytime? Maybe just do somthing like https://github.com/DataDog/datadog-agent/blob/42b8cfe9f821353592ad3c444a926a5eba706a13/pkg/collector/corechecks/servicediscovery/module/impl_linux.go#L384 or something in a module init() function so it's only done once?

File: pkg/collector/corechecks/servicediscovery/module/envs_test.go [hunk 7]
```diff
@@ -103,3 +105,67 @@ func memfile(name string, b []byte) (int, error) {
...
+	envMapCurr, err = getEnvsUsingScan(proc, true)
+	require.NoError(t, err)
+	require.Equal(t, envMapCurr, expectedMap)
+}
+
+func BenchmarkGetEnvs(b *testing.B) {
+	b.ResetTimer()
+	b.ReportAllocs()
+	for i := 0; i < b.N; i++ {
+		proc, err := customNewProcess(int32(os.Getpid()))
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/envs_test.go:147]: Maybe move the customNewProcess() call outside of the timed portion of the benchmark?

File: pkg/collector/corechecks/servicediscovery/module/envs_test.go [hunk 8]
```diff
@@ -103,3 +105,67 @@ func memfile(name string, b []byte) (int, error) {
...
+func TestEnvScanner(t *testing.T) {
+	curPid := os.Getpid()
+	proc, err := process.NewProcess(int32(curPid))
+	require.NoError(t, err)
+
+	envMapGood, err := getEnvs(proc)
+	require.NoError(t, err)
+
+	// extract environment variables using scanner
+	envMapCurr, err := getEnvsUsingScan(proc, false)
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/envs_test.go:118]: Why implement and test the false case if it's not used anywhere?

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 4 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: I've added all env. variables I've found in the code and unit tests

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 9]
```diff
@@ -153,3 +184,135 @@ func getEnvs(proc *process.Process) (map[string]string, error) {
...
+// otherwise collect all environment variables.
+type EnvScanner struct {
+	file    *os.File          // open pointer to environment variables file
+	scanner *bufio.Scanner    // scanner to read file data
+	targets map[uint64]string // map of environment variables of interest
+	envs    map[string]string // collected environment variables
+}
+
+// hashBytes return hash value of a bytes array using FNV-1a hash function
+func hashBytes(b []byte) uint64 {
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: Hash function allows to improve performance

File: pkg/collector/corechecks/servicediscovery/targetenvs/envscanner.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+package targetenvs
+
+import (
+	"bytes"
+	"fmt"
+	"hash/fnv"
+	"os"
+	"strconv"
+
+	"github.com/DataDog/datadog-agent/pkg/security/ptracer"
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/targetenvs/envscanner.go:15]: that's import caused an massive impact on the agent, and other component (https://github.com/DataDog/datadog-agent/pull/29574#issuecomment-2391880539)
why are you using this package?

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: do you have proof for that?

[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: yes, at some point I've benchmarked both:
targetMap[**uint64**]string
`BenchmarkGetEnvsTarget-4                 93546	     12868 ns/op	    6975 B/op	      41 allocs/op`
targetMap[**string**]{}
`BenchmarkGetEnvsTarget-4                 87042	     14122 ns/op	   10613 B/op	      89 allocs/op`

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 3 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: I refactored the code to add the `targetenvs` package, but still use exported string variables to have a single source of information regarding target environment variables.

File: pkg/collector/corechecks/servicediscovery/targetenvs/envscanner.go (hunk 10 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/targetenvs/envscanner.go:17]: Vincent suggested to import this package to use file scanner from it, I will remove import and create local code instead - it is just 30 lines of code

Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/targetenvs/envscanner.go]: I've removed imported ptracer in recent commit

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 5 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: I reverted `ptracer` package import due to impact, I copied function `zeroSplitter` from `ptracer` package

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: can you share the benchmak?

there is not doubt that using `map[uint64]string` is faster that `map[string]struct{}`, but you need to take into account the performance "penalty' you add once you calculate the hash of a string by yourself.
Meaning, you need to compare `map[string]struct{}` to `map[uint64]string + hash function`

[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: I was not clear, I used the same benchmark that measures function getTargetEnvs(proc), in my response above it was just title "targetMap" to distinguish two different results, I've not benchmarked the map itself.

[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: I removed usage of hash function from the code

File: pkg/collector/corechecks/servicediscovery/module/envs_test.go (hunk 8 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs_test.go]: I removed second parameter in this function

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+// Package envs provides target environment variables of interest.
+package envs
+
+import (
+	"fmt"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/servicediscovery/language"
+)
+
+// envKind separates environment variables by type, it helps to know when all target variables
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/envs/envs.go:15]: This optimization does not look correct. We can't know that just getting one `appEnv` is sufficient. Just to give one example the gunicorn parser falls back to `WSGI_APP` if there were problems in parsing `GUNICORN_CMD_ARGS`, so it's wrong to stop after getting `GUNICORN_CMD_ARGS`. I'd suggest you remove this optimization and just gather all the environment variables in the list.

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+	ddService                            // agent service env variable detected
+)
+
+type envLang struct {
+	kind envKind
+	lang language.Language
+}
+
+// targets is a collection of environment variables of interest.
+var targets = map[string]envLang{
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/envs/envs.go:33]: This should change to `map[string]struct{}` when the envLang stuff is removed.

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+	empty      envKind = 0
+	pwdEnv             = 1 << (iota - 1) // system settings env variable detected
+	appEnv                               // application system settings env variable detected
+	ddInjected                           // agent injection env variable detected
+	ddService                            // agent service env variable detected
+)
+
+type envLang struct {
+	kind envKind
+	lang language.Language
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/envs/envs.go:29]: `lang` is unused AFAICS.

File: pkg/collector/corechecks/servicediscovery/module/envs_test.go [hunk 14]
```diff
@@ -103,3 +106,49 @@ func memfile(name string, b []byte) (int, error) {
 
 	return fd, nil
 }
+
+// TestTargetEnvs it checks reading of target environment variables only from /proc/<pid>/environ.
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/module/envs_test.go:110]: AFAICS it doesn't test we don't read other variables than the ones we want.

File: pkg/collector/corechecks/servicediscovery/usm/python.go [hunk 15]
```diff
@@ -114,13 +112,13 @@ func (p pythonDetector) findNearestTopLevel(fp string) string {
 }
 
 func (g gunicornDetector) detect(args []string) (ServiceMetadata, bool) {
-	if fromEnv, ok := extractEnvVar(g.ctx.envs, gunicornEnvCmdArgs); ok {
+	if fromEnv, ok := extractEnvVar(g.ctx.envs, "GUNICORN_CMD_ARGS"); ok {
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/usm/python.go:115]: The changes in this file look unnecessary.

File: pkg/collector/corechecks/servicediscovery/usm/spring.go [hunk 16]
```diff
@@ -318,7 +318,7 @@ func (s springBootParser) GetSpringBootAppName(jarname string) (string, bool) {
 	combined := &props.Combined{Sources: []props.PropertyGetter{
 		newArgumentSource(s.ctx.args, "--"),
 		newArgumentSource(s.ctx.args, "-D"),
-		newEnvironmentSource(s.ctx.envs),
+		newEnvironmentSource(s.ctx.envs.GetVars()),
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/usm/spring.go:321]: This should not be using `GetVars`, the environment source should use the `Get` function from the type `EnvironmentVariables`, otherwise you don't get the benefit of having the test cases automatically catch cases where new variables are added to the parsing code but not added to the allowlist.

File: pkg/collector/corechecks/servicediscovery/apm/detect.go [hunk 17]
```diff
@@ -78,25 +93,10 @@ func isInjected(envs map[string]string) bool {
...
-	// in dd-trace-go are under 100 chars. The tracer.init example above at 51
-	// chars is one of the shortest.
-	ddTraceGoMaxLength = 100
-)
-
 // goDetector detects APM instrumentation for Go binaries by checking for
 // the presence of the dd-trace-go symbols in the ELF. This only works for
-// unstripped binaries.
-func goDetector(pid int, _ []string, _ map[string]string, _ usm.DetectorContextMap) Instrumentation {
+// non stripped binaries.
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/apm/detect.go:98]: It would be better to leave out unrelated changes like this.

File: pkg/collector/corechecks/servicediscovery/apm/detect.go [hunk 18]
```diff
@@ -37,7 +38,22 @@ const (
 	Injected Instrumentation = "injected"
 )
 
-type detector func(pid int, args []string, envs map[string]string, contextMap usm.DetectorContextMap) Instrumentation
+const (
```
Comment (vitkyrka) [inline pkg/collector/corechecks/servicediscovery/apm/detect.go:41]: Why has this code been moved?

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 11 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: I've removed search optimization in last commit

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 12 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: this struct has been changed in last commit

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 13 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: this struct has been removed

File: pkg/collector/corechecks/servicediscovery/module/envs_test.go (hunk 14 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs_test.go:109]: I've changed test to check that unexpected variables not present in result.

File: pkg/collector/corechecks/servicediscovery/usm/python.go (hunk 15 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/usm/python.go]: this code is restored to original

File: pkg/collector/corechecks/servicediscovery/usm/spring.go (hunk 16 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/usm/spring.go]: Ive changed it to use envs.Get()

File: pkg/collector/corechecks/servicediscovery/apm/detect.go (hunk 17 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/apm/detect.go]: I restored original comment

File: pkg/collector/corechecks/servicediscovery/apm/detect.go (hunk 18 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/apm/detect.go]: I've restored original code. Golang style expects all constants in the beginning of file.

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+	for env := range targets {
+		expectedEnvs = append(expectedEnvs, fmt.Sprintf("%s=true", env))
+		expectedMap[env] = "true"
+	}
+	return expectedEnvs, expectedMap
+}
+
+// EnvironmentVariables - collected of targeted environment variables.
+type EnvironmentVariables struct {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/envs/envs.go:48]: neat: Following clean go, you can rename the struct (and the functions) to be `Variables`
as the usage outside the packages will be `envs.Variables` instead of `envs.EnvironmentVariables`

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+	"SPRING_APPLICATION_NAME":  {},
+	"SPRING_CONFIG_LOCATIONS":  {},
+	"SPRING_CONFIG_NAME":       {},
+	"SPRING_PROFILES_ACTIVE":   {},
+}
+
+// GetExpectedEnvs - return list of expected env. variables for testing.
+func GetExpectedEnvs() ([]string, map[string]string) {
+	var expectedEnvs []string
+	var expectedMap = make(map[string]string)
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/envs/envs.go:38]: neat - preallocated maps and arrays are always preferable (performance wise)

```suggestion
	expectedEnvs := make([]string, 0, len(targets))
	expectedMap := make(map[string]string, len(targets))
```

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+	"JAVA_OPTIONS":             {},
+	"JDPA_OPTS":                {},
+	"SPRING_APPLICATION_NAME":  {},
+	"SPRING_CONFIG_LOCATIONS":  {},
+	"SPRING_CONFIG_NAME":       {},
+	"SPRING_PROFILES_ACTIVE":   {},
+}
+
+// GetExpectedEnvs - return list of expected env. variables for testing.
+func GetExpectedEnvs() ([]string, map[string]string) {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/envs/envs.go:36]: if that's only for testing, let's move it to a dedicated file (envs_testutils.go) and add a build flag `//go:build test`

it will allows us ensure no-testing only code base enters to our production versions

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+		return "", false
+	}
+
+	val, ok := ev.vars[name]
+	return val, ok
+}
+
+// Set saves the environment variable if it is targeted.
+// returns true if env variable matches the target
+func (ev *EnvironmentVariables) Set(name string, val string) bool {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/envs/envs.go:76]:
```suggestion
func (ev *EnvironmentVariables) Set(name, val string) bool {
```

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+// Get returns an environment variable if it is present in the collection
+func (ev *EnvironmentVariables) Get(name string) (string, bool) {
+	if _, ok := targets[name]; !ok {
+		return "", false
+	}
+
+	val, ok := ev.vars[name]
+	return val, ok
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/envs/envs.go:71]: imho, `vars` should reflect only the allowed environment variables
no need to store (and waste memory) unneeded envs + currently you always pay for 2 map accesses per "supported" environment variable instead of only one map access

Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go:71]: If `vars` is guaranteed to have only supported environment variables, then no need to check `targets` first

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+func (ev *EnvironmentVariables) Set(name string, val string) bool {
+	if _, ok := targets[name]; !ok {
+		return false
+	}
+	if ev.vars == nil {
+		ev.vars = make(map[string]string)
+	}
+	ev.vars[name] = val
+
+	return false
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/envs/envs.go:85]:
```suggestion
	return true
```

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 25]
```diff
@@ -153,3 +156,86 @@ func getEnvs(proc *process.Process) (map[string]string, error) {
...
+	scanner.Split(zeroSplitter)
+
+	return &EnvReader{
+		file:    file,
+		scanner: scanner,
+		envs:    envs.NewEnvironmentVariables(nil),
+	}, nil
+}
+
+// finish closes an open file.
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/module/envs.go:197]: neat: rename to Close

File: pkg/collector/corechecks/servicediscovery/usm/python_test.go [hunk 26]
```diff
@@ -10,6 +10,7 @@ import (
 	"testing"
 	"testing/fstest"
 
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/servicediscovery/envs"
 	"github.com/stretchr/testify/require"
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/usm/python_test.go:14]:
```suggestion
	"github.com/stretchr/testify/require"

	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/servicediscovery/envs"
```

File: pkg/collector/corechecks/servicediscovery/usm/spring.go [hunk 27]
```diff
@@ -352,6 +352,23 @@ func (s springBootParser) GetSpringBootAppName(jarname string) (string, bool) {
...
 
+var springEnvs = []string{
+	"SPRING_APPLICATION_NAME",
+	"SPRING_CONFIG_LOCATIONS",
+	"SPRING_CONFIG_NAME",
+	"SPRING_PROFILES_ACTIVE",
+}
+
+func (s springBootParser) getSpringEnvs() map[string]string {
+	envsMap := make(map[string]string)
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/usm/spring.go:363]: prealloc the maps 
```suggestion
	envsMap := make(map[string]string, len(springEnvs))
```

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 19 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: I've renamed this struct in recent commit

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 20 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: done

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 21 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: I've created separate file envs_testutils.go with a code dedicated for tests only

File: pkg/collector/corechecks/servicediscovery/module/envs.go (hunk 25 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: done

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 24 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: fixed

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 23 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: This is where unnecessary environment variables are filtered out and ultimately only the requested variables are included in the list.

File: pkg/collector/corechecks/servicediscovery/usm/python_test.go (hunk 26 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/usm/python_test.go]: fixed

File: pkg/collector/corechecks/servicediscovery/usm/spring.go (hunk 27 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/usm/spring.go]: done

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+	if _, ok := targets[name]; !ok {
+		return "", false
+	}
+
+	val, ok := ev.vars[name]
+	return val, ok
+}
+
+// Set saves the environment variable if it is targeted.
+// returns true if env variable matches the target
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go:54]: I've fixed this function

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 22 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: done

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 29]
```diff
@@ -153,3 +156,86 @@ func getEnvs(proc *process.Process) (map[string]string, error) {
...
+	if found {
+		return er.envs.Set(name, val)
+	}
+	return false
+}
+
+// getTargetEnvs reads the environment variables of interest from the /proc/<pid>/environ file.
+func getTargetEnvs(proc *process.Process) (envs.EnvironmentVariables, error) {
+	er, err := newEnvReader(proc)
+	defer er.finish()
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: fixed

File: pkg/collector/corechecks/servicediscovery/module/envs.go [hunk 30]
```diff
@@ -153,3 +156,86 @@ func getEnvs(proc *process.Process) (map[string]string, error) {
...
+
+// finish closes an open file.
+func (er *EnvReader) finish() {
+	if er.file != nil {
+		er.file.Close()
+	}
+}
+
+// add adds env. variable to the map of environment variables,
+// returns true if reading should be stopped.
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/module/envs.go]: I changed function and removed comment

File: pkg/collector/corechecks/servicediscovery/usm/spring.go [hunk 31]
```diff
@@ -352,6 +352,23 @@ func (s springBootParser) GetSpringBootAppName(jarname string) (string, bool) {
 	return conf.Get(appnamePropName)
 }
 
+var springEnvs = []string{
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/usm/spring.go]: I've removed this var

File: pkg/collector/corechecks/servicediscovery/envs/envs.go [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+	if _, ok := targets[name]; !ok {
+		return "", false
+	}
+
+	val, ok := ev.Vars[name]
+	return val, ok
+}
+
+// GetDefault returns an environment variable or provided default.
+func (ev *Variables) GetDefault(name string, defVal string) string {
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/envs/envs.go:54]:
```suggestion
func (ev *Variables) GetDefault(name, defVal string) string {
```

File: pkg/collector/corechecks/servicediscovery/envs/envs_testutils.go [hunk 33]
```diff
@@ -0,0 +1,30 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build test || bench
```
Comment (guyarb) [inline pkg/collector/corechecks/servicediscovery/envs/envs_testutils.go:6]: bench is not a know buildflag

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 23 above)
Comment (guyarb) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: I don't get the code then, and how you spare redundant memory

We have 100 environment variable for process X, we care about only 10 of them
We store the entire 100 environment variables in the memory, but we only allow accessing the 10?

The filtering should be while reading the envs or when creating the `Variables`
We should not read redundant environment variable, we should not store redundant environment variable, and we should not allow the consumers of `envs.Variables` to access redundant environment variables

As far as I understand the code, only the third requirement is fulfilled.
We should at least fulfill the second requirement as well

Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: The Set() command is called while reading file. Flow is Scan() -> returns next string -> add() -> Set() -> check if need it -> Scan(next). Only target variables are stored.

Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: Expanding my previous answer. The file scanner reads "environ" file string by string and returns true when next string was read. The "add" function fetches current string and passes env variable name and value to envs.Set, this function saves only targeted variable and scanner goes to the next string in file. If variable is not of interest it is discarded and not stored in any place.

Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: What is perhaps not very clear is that the check on `Get` is for test cases. In the test cases the map passed in to `NewVariables` could have variables which are not on the allowlist (in case new code is being added), so for this check on `Get` prevents these variables from being used until they have been added to the `allowlist`, to avoid missing new variables. The check on `Get` could be dropped if the `NewVariables` function is made to do the check for valid variables instead. This will only affect the tests, since in the normal case nil or an empty mapped will be passed in to NewVariables() since variables are added as they are read.

Comment (vitkyrka) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: I'd suggest the above change to move the check to NewVariables be done in a follow-up PR.

**Review by vitkyrka — APPROVED** [review]

File: pkg/collector/corechecks/servicediscovery/envs/envs.go (hunk 23 above)
Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: Ok, I'll modify Get() function and will move target verification to stage of struct initialization in a follow-up PR

Comment (yuri-lipnesh) [reply pkg/collector/corechecks/servicediscovery/envs/envs.go]: this is follow-up [PR ](https://github.com/DataDog/datadog-agent/pull/30068)
